File: flink-frame/src/main/java/com/yg/flink/frame/KafkaApp.scala
package com.yg.flink.frame

import java.util.Properties

import com.google.gson.{JsonObject, JsonParser}
import com.yg.flink.frame.conf.ApiHelper
import org.apache.flink.api.common.serialization.SimpleStringSchema
import org.apache.flink.api.scala._
import org.apache.flink.core.fs.Path
import org.apache.flink.formats.parquet.avro.ParquetAvroWriters
import org.apache.flink.streaming.api.functions.sink.filesystem.rollingpolicies.OnCheckpointRollingPolicy
import org.apache.flink.streaming.api.functions.sink.filesystem.{OutputFileConfig, StreamingFileSink}
import org.apache.flink.streaming.api.scala.{DataStream, StreamExecutionEnvironment}
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer

/**
 * TODO
 *
 * @author Y.G
 * @date 2020/5/7 16:05
 **/
object KafkaApp {

  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment

    // 每 1000ms 开始一次 checkpoint
    env.enableCheckpointing(10 * 60 * 1000L)
    val prop = new Properties()
    prop.setProperty("bootstrap.servers", ApiHelper.loadConf("opt.source.bootstrap.servers"))
    prop.setProperty("group.id", "yg_kafka_1")
    val kafkaStream: DataStream[String] = env.addSource(new FlinkKafkaConsumer[String](ApiHelper.loadConf("opt.source.topic"), new SimpleStringSchema(), prop))
    kafkaStream.print()

    val builder = StreamingFileSink
      .forBulkFormat(new Path(ApiHelper.loadConf("opt.sink.path")), ParquetAvroWriters.forReflectRecord(classOf[DbMovie]))
    builder.withRollingPolicy(OnCheckpointRollingPolicy.build())
      .withOutputFileConfig(OutputFileConfig.builder()
        .withPartPrefix("yg")
        .withPartSuffix(".parquet")
        .build())

    val parquetSink: StreamingFileSink[DbMovie] = builder.build()
    kafkaStream
      .map(v => {
        val jsonObject: JsonObject = JsonParser.parseString(v).getAsJsonObject
        DbMovie(jsonObject.get("cover").getAsString, jsonObject.get("title").getAsString)
      })
      .filter(k => k != null)
      .addSink(parquetSink)

    env.execute("kafka_yg")
  }

  case class DbMovie(cover:String, title:String)
}

File: flink-demo/src/main/java/com/yg/flink/demo/KafkaApp.java
package com.yg.flink.demo;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.yg.flink.demo.entity.DbMovie;
import org.apache.flink.api.common.serialization.SimpleStringSchema;
import org.apache.flink.core.fs.Path;
import org.apache.flink.formats.parquet.avro.ParquetAvroWriters;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.sink.filesystem.StreamingFileSink;
import org.apache.flink.streaming.api.functions.sink.filesystem.rollingpolicies.OnCheckpointRollingPolicy;
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer;

import java.util.Properties;

/**
 * TODO
 *
 * @author Y.G
 * @date 2020/5/9 11:18
 **/
public class KafkaApp {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.enableCheckpointing(10*60*1000L);
        Properties prop = new Properties();
        prop.setProperty("bootstrap.servers", "192.168.90.96:9092");
        prop.setProperty("group.id", "yg_test3");
        DataStreamSource<String> kafkaStream = env.addSource(new FlinkKafkaConsumer<>("db_movie", new SimpleStringSchema(), prop));
        kafkaStream.print();

        // parquet Sink

        StreamingFileSink<DbMovie> sFileSink = StreamingFileSink
                .forBulkFormat(new Path("hdfs://myha/tmp/yg/flink"), ParquetAvroWriters.forReflectRecord(DbMovie.class))
                .withRollingPolicy(OnCheckpointRollingPolicy.build())
                .build();

        kafkaStream.map(v -> {
            JsonObject obj = JsonParser.parseString(v).getAsJsonObject();
            DbMovie dbMovie = new DbMovie();
            String cover = obj.get("cover").getAsString();
            String title = obj.get("title").getAsString();
            dbMovie.setCover(cover == null ? "" : cover);
            dbMovie.setTitle(title == null ? "" : title);
            return dbMovie;
        }).addSink(sFileSink);

        env.execute("db_movie");
    }
}
